Update linter version and fix all the errors

earlgreyz opened the pull request:
This commit is extremely huge but the only changes are full stops at the end of comments and new lines here and there. Stacked on top of https://github.com/facebookincubator/fbender/pull/16.

As for the changes in `.golangci.yml`:
* `gomnd` is not very useful, as it makes you create constant out of every number such as in `2*x` you need to create a constant for 2....
* `exhaustivestruct` requires to fill all the struct fields when creating a struct, which would add a lot of `nil` fields where not needed
*  `goerr113` and `wrapcheck` are reasonable but it require to change the logic in some places a little bit so I will fix it in a separate diff
* `paralleltest` will require some code changes so I will fix it in a separate diff

fabriziocucci commented:
@earlgreyz thanks for putting this up! 🥇

My knowledge about this project is quite limited so apologies in advance if this might look like an obvious question.

A few weeks back, we dropped Travis in favour of GitHub Actions so now I'm wondering: do we really need CircleCI at all?

earlgreyz commented:
@fabriziocucci it was separated from Travis to have a separation between builds failing because of linting and builds failing due to actual test errors. I think it can be moved into another GitHub action to replace CircleCI. I will put it as a separate commit in this PR.

earlgreyz commented:
CircleCI is going to fail as I removed its config, but it's expected. It will have to get manually removed from the hooks after this is merged. I have copied and modified the workflow for build, but maybe linters do not need to run on both 1.14 and 1.15 and just run them on 1.15?

fabriziocucci commented:
Thanks for the explanation @earlgreyz! 👍 

If you find yourself replicating a lot of logic from the existing GitHub Action, it could make sense to simply put it there as an additional step. IMHO, it's were it should be if formatting is something that we really want to enforce anyway.

fabriziocucci commented on .github/workflows/lint.yml:
```diff
@@ -0,0 +1,32 @@
+    - name: Get dependencies
+      run: go mod download
+
+    - name: Download Linter
+      run: curl -sSfL https://raw.githubusercontent.com/golangci/golangci-lint/master/install.sh | sh -s -- -b $(go env GOPATH)/bin v1.34.1
+
+    - name: Lint
+      run: golangci-lint run
```
Why not simply adding these to the main `build.yml` befre build and test steps?

Seems a bit of a waste replicating the whole GitHub Action! 😕

earlgreyz replied:
This will cause the tests not to run at all if linting fails. As far as I can remember this was the reason to separate those two, so in case when linting fails the tests and code coverage are still checked.

fabriziocucci replied:
Yeah, I'm questioning the rationale because I've seen other projects where the lint is just a step of the build.

If the only concern is not failing the whole job, we could set `continue-on-error: true` (see [GitHub Action doc](https://docs.github.com/en/free-pro-team@latest/actions/reference/workflow-syntax-for-github-actions#jobsjob_idstepscontinue-on-error)).

What do you think?

earlgreyz replied:
As far as I understand `continue-on-error: true` will make the whole job pass despite lint errors, which I don't think is what we want. In this case I think it's better to make it a step of the build. The only benefit of having it as a separate job is that than in the PR you can immediately see:

- Tests passed
- Lint passed
- Coverage check passed (btw I have noticed it's not here, maybe the token env variable has not been set when migrating to GitHub actions? if you could take a look)

as opposed to 
- Tests **and** lint passed
- Coverage check passed

so when making a PR with both tests failing and lint errors you can see that and fix both before pushing a new version instead of first fixing tests/lint (depending on which one is defined first in the job), pushing again just to see that the other one has also failed.

However since the repo is not huge and tests take only few minutes to run this is not a big issue and. So let me know if you think it's better to have it in one job and I'll put it after the tests in the main job without `continue-on-error` flag.

fabriziocucci replied:
> As far as I understand continue-on-error: true will make the whole job pass despite lint errors, which I don't think is what we want. 

I was hoping we could see at least a warning or additional indicator next to the green tick in case of lint failure but apparently that's not the case (see [feature discussion](https://github.community/t/continue-on-error-allow-failure-ui-indication/16773)).

Let's keep it separate then! :)

fabriziocucci commented:
![](https://media1.giphy.com/media/qgri3D9sTwCUGMcT8L/200.gif?cid=5a38a5a27t2sym9n3jm38plyh875uitkazu85ymj9uiwxchk&rid=200.gif)

Thanks @earlgreyz! 👏
